gcloud-java-contrib/gcloud-java-nio/src/main/java/com/google/gcloud/storage/contrib/nio/CloudStorageFileSystemProvider.java

```diff
@@ -421,13 +478,24 @@ public void createDirectory(Path dir, FileAttribute<?>... attrs) {
   }
 
   /**
-   * This feature is not supported. Please use {@link #readAttributes(Path, Class, LinkOption...)}
-   *
-   * @throws UnsupportedOperationException
    */
   @Override
   public Map<String, Object> readAttributes(Path path, String attributes, LinkOption... options) {
-    throw new UnsupportedOperationException();
+    CloudStoragePath cloudPath = checkPath(path);
+    checkNotNull(attributes);
+    BlobInfo blobInfo = gcsStorage.get(cloudPath.getBlobId());
+    HashMap<String, Object> ret = new HashMap<>();
+    Map<String, String> metadata = blobInfo.metadata();
+    // it's faster to copy everything than to parse the attributes string.
+    if (null != metadata) {
+      for (Map.Entry<String, String> entry : metadata.entrySet()) {
+        ret.put(entry.getKey(), entry.getValue());
+      }
+    }
+    ret.put("content-type", blobInfo.contentType());
+    ret.put("cache-control", blobInfo.cacheControl());
+    ret.put("generation", blobInfo.generation());
+    return ret;
```

Comment:
The tests passed before without this method being implemented. Did you add code to the tests that necessitated the implementation of this method? If so, could the tests be modified so that this method does not need to be implemented? I ask because implementing this method is adding a feature, which is arguably out of the scope of this PR. It might be better if this functionality were added later in its own PR.